examples/src/test/scala/docs/guide/ShoppingCartApp.scala

```diff
@@ -0,0 +1,270 @@
+  def deleteCartState(cartId: String): Future[Done]
+}
+
+class DailyCheckoutProjectionRepositoryImpl(session: CassandraSession)(implicit val ec: ExecutionContext)
+    extends DailyCheckoutProjectionRepository {
+  import DailyCheckoutProjectionHandler._
+
+  val keyspace = "akka_projection"
+  val dailyCheckoutsTable = "daily_checkouts"
+  val cartStateTable = "cart_state"
+
+  def checkoutCountsForDate(date: LocalDate): Future[Seq[Checkout]] = {
+    session
+      .selectAll(s"SELECT cart_id, item_id, quantity FROM $keyspace.$dailyCheckoutsTable WHERE date = ?", date)
+      .map { rows =>
+        rows.map(row => Checkout(date, row.getString("cart_id"), row.getString("item_id"), row.getInt("quantity")))
+      }
+  }
+
+  def addCheckout(checkoutItem: Checkout): Future[Done] = {
+    import checkoutItem._
+    session.executeWrite(
+      s"INSERT INTO $keyspace.$dailyCheckoutsTable (date, cart_id, item_id, quantity) VALUES (?, ?, ?, ?)",
+      date,
+      cartId,
+      itemId,
+      Int.box(quantity))
+  }
+
+  def cartState(): Future[Seq[CartState]] = {
+    session
+      .selectAll(s"SELECT cart_id, item_id, quantity FROM $keyspace.$cartStateTable")
+      .map { rows =>
+        rows.map(row => CartState(row.getString("cart_id"), row.getString("item_id"), row.getInt("quantity")))
+      }
+  }
+
+  def updateCartState(cartState: CartState): Future[Done] = {
+    import cartState._
+    session.executeWrite(
+      s"UPDATE $keyspace.$cartStateTable SET quantity = ? WHERE cart_id = ? AND item_id = ?",
+      Integer.valueOf(quantity),
+      cartId,
+      itemId)
+  }
+
+  def deleteCartState(cartId: String): Future[Done] = {
+    session.executeWrite(s"DELETE FROM $keyspace.$cartStateTable WHERE cart_id = ?", cartId)
+  }
+}
+//#guideProjectionRepo
+
+//#guideProjectionHandler
+object DailyCheckoutProjectionHandler {
+  final case class Checkout(date: LocalDate, cartId: String, itemId: String, quantity: Int)
+  final case class CartState(cartId: String, itemId: String, quantity: Int)
+
+  def toDate(date: Instant) = LocalDate.from(date.atZone(ZoneId.of("UTC")))
+}
+
+class DailyCheckoutProjectionHandler(tag: String, system: ActorSystem[_], repo: DailyCheckoutProjectionRepository)
+    extends StatefulHandler[Map[String, Map[String, CartState]], EventEnvelope[ShoppingCartEvents.Event]]()(
+      system.classicSystem.dispatcher) {
+
+  import DailyCheckoutProjectionHandler._
+
+  @volatile private var checkoutCounter = 0
+  private val log = LoggerFactory.getLogger(getClass)
+  private implicit val ec: ExecutionContext = system.classicSystem.dispatcher
+
+  def initialState(): Future[Map[String, Map[String, CartState]]] =
+    repo.cartState().map { rows =>
+      val byCartId = rows.groupBy(_.cartId)
+      val byCartAndItem = byCartId.view.map {
+        case (k, v) => k -> v.groupBy(_.itemId).map { case (itemId, states) => (itemId -> states.head) }
+      }.toMap
+      byCartAndItem
+    }
+
```

Comment:
This looks like it's keeping a full memory image of the shopping carts. I'm not sure I see the purpose of a StatefulHandler for this guide. 

It may be simpler to start with a stateless one?

Or maybe indicate why stateful is useful in this case...

Reply:
Oh, I think I see it now. Events are increments so applying each event requires the handler to be stateful.

Do we have some mechanism to keep the in-memory state idempotent, then?

Reply:
#350